engine/kv_engine.cpp

```diff
@@ -1720,3 +1736,193 @@ Status KVEngine::RestoreDlistRecords(DLRecord *pmp_record) {
 }
 
 } // namespace KVDK_NAMESPACE
+
+namespace KVDK_NAMESPACE {
+std::unique_ptr<Queue> KVEngine::createQueue(StringView const collection_name) {
+  std::uint64_t ts = get_timestamp();
+  uint64_t id = list_id_.fetch_add(1);
+  std::string name(collection_name.data(), collection_name.size());
+  std::unique_ptr<Queue> queue_uptr =
+      std::make_unique<Queue>(pmem_allocator_.get(), name, id, ts);
+  return queue_uptr;
+}
+
+Queue *KVEngine::findQueue(StringView const collection_name) {
+  HashTable::KeyHashHint hint = hash_table_->GetHint(collection_name);
+  HashEntry hash_entry;
+  HashEntry *entry_ptr = nullptr;
+  Status s =
+      hash_table_->SearchForRead(hint, collection_name, RecordType::QueueRecord,
+                                 &entry_ptr, &hash_entry, nullptr);
+  switch (s) {
+  case Status::NotFound: {
+    return nullptr;
+  }
+  case Status::Ok: {
+    return hash_entry.queue_ptr;
+  }
+  default: {
+    kvdk_assert(false, "Invalid state in findQueue()!");
+    return nullptr;
+  }
+  }
+}
+
+Status KVEngine::xPop(StringView const collection_name, std::string *value,
+                      KVEngine::QueueOpWhere where) {
+  Status s = MaybeInitWriteThread();
+  if (s != Status::Ok) {
+    return s;
+  }
+  Queue *queue_ptr = findQueue(collection_name);
+  if (!queue_ptr) {
+    return Status::NotFound;
+  }
+  bool pop_success = false;
+  switch (where) {
+  case QueueOpWhere::Left: {
+    pop_success = queue_ptr->PopFront(value);
+    break;
+  }
+  case QueueOpWhere::Right: {
+    pop_success = queue_ptr->PopBack(value);
+    break;
+  }
+  default: {
+    GlobalLogger.Error("Impossible!");
+    return Status::Abort;
+  }
+  }
+  if (pop_success) {
+    return Status::Ok;
+  } else {
+    return Status::NotFound;
+  }
+}
+
+Status KVEngine::xPush(StringView const collection_name, StringView const value,
+                       KVEngine::QueueOpWhere where) try {
+  Status s = MaybeInitWriteThread();
+  if (s != Status::Ok) {
+    return s;
+  }
+
+  Queue *queue_ptr = nullptr;
+
+  // Find UnorederedCollection, create if none exists
+  {
+    queue_ptr = findQueue(collection_name);
+    if (!queue_ptr) {
+      HashTable::KeyHashHint hint_collection =
+          hash_table_->GetHint(collection_name);
+      std::unique_lock<SpinMutex> lock_collection{*hint_collection.spin};
+      {
+        // Lock and find again in case other threads have created the
+        // UnorderedCollection
+        queue_ptr = findQueue(collection_name);
+        if (queue_ptr) {
```

Comment:
maybe can simply write like this, `if(!queue_ptr){.....}`.

Reply:
Not so consistent with other code snippets. Also not so friendly for debugging.

Reply:
simply is clear. Why you think not friendly for debugging?